File: app/discover/api_fetch_end_points.rb
# fetch the end points for a given project (tenant)
# return list of regions, to allow further recursive scanning

require_relative 'api_access'
require_relative 'inventory_mgr'
require_relative 'configuration'
require_relative 'util'
require 'json'

class ApiFetchEndPoints < ApiAccess
  include Util
  
  def initialize()
    @inventory = InventoryMgr.instance
  end
  
  def get(id)
    if id != "admin"
      return [] # XXX currently having problems authenticating to other tenants
    end
    v2_auth_pwd(id)
    
    environment = Configuration.instance.get_env()
    regions = []
    services = @@body_hash['access']['serviceCatalog']
    services.each {|s|
      (s["type"] != "identity") && next
      e = s["endpoints"][0]
      e[:environment] = environment
      e[:project] = id
      e[:type] = "endpoint"
      symbolize_keys_deep!(e)
      @inventory.set(e)
      region = nil
      region = {
        :id => e[:region],
        :environment => environment,
        :text => e[:region],
        :descendants => 3,
        :parent_id => environment,
        :parent_type => "environment"
      }
      regions.push(region)
    }
    
    return regions
  end
  
end  

File: app/discover/api_fetch_end_points.py
# fetch the end points for a given project (tenant)
# return list of regions, to allow further recursive scanning

from discover.api_access import ApiAccess


class ApiFetchEndPoints(ApiAccess):

    def get(self, project_id):
        if project_id != "admin":
            return []  # XXX currently having problems authenticating to other tenants
        self.v2_auth_pwd(project_id)

        environment = ApiAccess.config.get_env_name()
        regions = []
        services = ApiAccess.auth_response['access']['serviceCatalog']
        endpoints = []
        for s in services:
            if s["type"] != "identity":
                continue
            e = s["endpoints"][0]
            e["environment"] = environment
            e["project"] = project_id
            e["type"] = "endpoint"
            endpoints.append(e)
        return endpoints
